File: packages/core-dart/lib/src/routing/memo.dart
import '../address/codes.dart';
import 'safe_routing_id.dart';

class NormalizeResult {
  final String? normalized;
  final List<Warning> warnings;

  NormalizeResult({this.normalized, required this.warnings});
}

final BigInt uint64Max = SafeRoutingId.uint64Max;
final RegExp digitsOnly = RegExp(r'^\d+$');

/// Strict normalizer for MEMO_ID type.
/// A MEMO_ID must be a non-empty string of digits parseable as a uint64.
/// Leading zeros are invalid (except the canonical "0").
/// Returns null if the value cannot be used as a routing ID.
///
/// Web safety: range validation goes through [SafeRoutingId.tryParse],
/// which works on the decimal string directly (length- and
/// lexicographic-comparison based) and therefore never coerces the value
/// through `int`/JS `Number`. Large MEMO_IDs (>
/// `Number.MAX_SAFE_INTEGER`) are validated exactly in browser contexts.
NormalizeResult normalizeMemoId(String s) {
  final warnings = <Warning>[];

  // Reject blank or non-digit strings
  if (s.isEmpty || !digitsOnly.hasMatch(s)) {
    return NormalizeResult(normalized: null, warnings: warnings);
  }

  // Reject leading zeros (e.g. "007" is invalid for a strict MEMO_ID)
  if (s.length > 1 && s.startsWith('0')) {
    warnings.add(
      Warning(
        code: WarningCode.nonCanonicalRoutingId,
        severity: 'warn',
        message:
            'Memo routing ID had leading zeros. Normalized to canonical decimal.',
        normalization: Normalization(
          original: s,
          normalized: BigInt.parse(s).toString(),
        ),
      ),
    );
    // Strip zeros and re-normalize for the returned value
    final stripped = BigInt.parse(s).toString();
    if (SafeRoutingId.tryParseStrict(stripped) == null) {
      return NormalizeResult(normalized: null, warnings: warnings);
    }
    return NormalizeResult(normalized: stripped, warnings: warnings);
  }

  // Validate uint64 range (string-exact; safe on Flutter Web)
  if (SafeRoutingId.tryParseStrict(s) == null) {
    return NormalizeResult(normalized: null, warnings: warnings);
  }

  return NormalizeResult(normalized: s, warnings: warnings);
}

/// Normalizer for MEMO_TEXT type — tries to parse a numeric routing ID.
/// Leading zeros trigger a normalization warning; non-numeric values return null.
///
/// Cross-SDK contract (identical in TypeScript `normalizeMemoTextId` and Go
/// `NormalizeMemoTextID`, enforced by `spec/memo_text_normalization.json`):
/// - The input is never trimmed. Surrounding or embedded whitespace
///   (e.g. `' 123 '`) makes the memo unroutable: `normalized` is null and no
///   warning is emitted.
/// - Only ASCII digits `0-9` are accepted (no sign, exponent, separators or
///   non-ASCII digits).
/// - Leading zeros (e.g. `'00123'`) are stripped to the canonical form
///   (`'123'`, or `'0'` for all zeros) and a single `NON_CANONICAL_ROUTING_ID`
///   warning is emitted with `normalization.original` set to the raw input and
///   `normalization.normalized` set to the stripped value.
/// - Values above uint64 max are unroutable; any leading-zero warning is kept.
NormalizeResult normalizeMemoTextId(String s) {
  final warnings = <Warning>[];

  // Step 1, 2, 3: Blank, non-digit
  if (s.isEmpty || !digitsOnly.hasMatch(s)) {
    return NormalizeResult(normalized: null, warnings: warnings);
  }

  // Step 4: Leading zeros — normalize and warn
  var normalized = s.replaceFirst(RegExp(r'^0+'), '');
  if (normalized.isEmpty) {
    normalized = '0';
  }

  if (normalized != s) {
    warnings.add(
      Warning(
        code: WarningCode.nonCanonicalRoutingId,
        severity: 'warn',
        message:
            'Memo routing ID had leading zeros. Normalized to canonical decimal.',
        normalization: Normalization(original: s, normalized: normalized),
      ),
    );
  }

  // Step 5: uint64 max — validated on the string itself, exactly, so IDs
  // above JS Number.MAX_SAFE_INTEGER are not truncated on Flutter Web.
  if (SafeRoutingId.tryParse(normalized) == null) {
    return NormalizeResult(normalized: null, warnings: warnings);
  }

  return NormalizeResult(normalized: normalized, warnings: warnings);
}

File: packages/core-ts/src/routing/memo.ts
import { Warning } from "../address/types";

export type NormalizeResult = {
  normalized: string | null;
  warnings: Warning[];
};

/**
 * Maximum value for a 64-bit unsigned integer (uint64).
 */
const UINT64_MAX = BigInt("18446744073709551615");

/**
 * Normalizes a numeric string into a canonical uint64 representation.
 * Strips leading zeros and validates that the value is within uint64 boundaries.
 *
 * Cross-SDK contract (identical in Go `NormalizeMemoTextID` and Dart
 * `normalizeMemoTextId`, enforced by `spec/memo_text_normalization.json`):
 * - The input is never trimmed. Surrounding or embedded whitespace
 *   (e.g. `" 123 "`) makes the memo unroutable: `normalized` is null and no
 *   warning is emitted.
 * - Only ASCII digits `0-9` are accepted (no sign, exponent, separators or
 *   non-ASCII digits).
 * - Leading zeros (e.g. `"00123"`) are stripped to the canonical form
 *   (`"123"`, or `"0"` for all zeros) and a single `NON_CANONICAL_ROUTING_ID`
 *   warning is emitted with `normalization.original` set to the raw input and
 *   `normalization.normalized` set to the stripped value.
 * - Values above uint64 max are unroutable; any leading-zero warning is kept.
 *

 * @param s - The numeric string to normalize.
 * @returns Result containing the normalized string (or null if invalid) and any warnings.
 */
export function normalizeMemoTextId(s: string): NormalizeResult {
  const warnings: Warning[] = [];

  if (s.length === 0 || !/^\d+$/.test(s)) {
    return { normalized: null, warnings };
  }

  let normalized = s.replace(/^0+/, "");
  if (normalized === "") {
    normalized = "0";
  }

  if (normalized !== s) {
    warnings.push({
      code: "NON_CANONICAL_ROUTING_ID",
      severity: "warn",
      message:
        "Memo routing ID had leading zeros. Normalized to canonical decimal.",
      normalization: { original: s, normalized },
    });
  }

  // Validate that normalized contains only decimal digits before BigInt conversion
  if (!/^\d+$/.test(normalized)) {
    return { normalized: null, warnings };
  }

  try {
    const val = BigInt(normalized);
    if (val > UINT64_MAX) {
      return { normalized: null, warnings };
    }
  } catch {
    return { normalized: null, warnings };
  }

  return { normalized, warnings };
}
